### PR TITLE
service/raft: specialized verb for failure detector pinger

### PR DESCRIPTION
**Pull request by kbr-**
We used GOSSIP_ECHO verb to perform failure detection. Now we use
a special verb DIRECT_FD_PING introduced for this purpose.

There are multiple reasons to do so.

One minor reason: we want to use the same connection as other Raft
verbs: if we can't deliver Raft append_entries or vote messages
somewhere, that endpoint should be marked dead; if we can, the
endpoint should be marked alive. So putting pings on the same
connection as the other Raft verbs is important when dealing with
weird situations where some connections are available but others are
not. Observe that in `do_get_rpc_client_idx`, we put the new verb in
the right place.

Another minor reason: we remove the awkward gossiper `echo_pinger`
abstraction which required storing and updating gossiper generation
numbers. This also removes one dependency from Raft service code to
gossiper.

Major reason 1: the gossip echo handler has a weird mechanism where a
replacing node returns errors during the replace operation to some of
the nodes. In Raft however, we want to mark servers as alive when they
are alive, including a server running on a node that's replacing
another node.

Major reason 2, related to the previous one: when server B is
replacing server A with the same IP, the failure detector will try to
ping both servers. Both servers are mapped to the same IP by the
address map, so pings to both servers will reach server B. We want
server B to respond to the pings destined for server B, but not to
pings destined for server A, so the sender can mark B alive but keep A
marked dead.

To do this, we include the destination's Raft ID in our RPCs. The
destination compares the received ID with its own. If it's different,
it returns a `wrong_destination` response, and the failure detector
knows that the ping did not reach the destination (it reached someone
else).

Yet another reason: removes "Not ready to respond gossip echo
message" log spam during replace.

### CODE REVIEW

**Comment by kostja** [comment]
Will gossip have to keep echo messages forever after this change?

**Comment by kbr-** [comment]
> Will gossip have to keep echo messages forever after this change?

Can't answer at this point.

Right now gossiper's failure detection  has a bit different requirements than Raft's failure detection. For example, because of reasons (refs: 40b60e8f09fda83248cba44374821a996ebdb036, #8013), gossiper wants to keep the replacing node marked as DOWN during the replace operation, which is done by the echo message handler returning an exception on the replacing node. On the other hand, we want to keep it simple correctly detect servers if they are dead or alive when they join group 0, without special treatment for some servers and not others.

I can guess that in the future when we migrate all topology management to group 0, the problems solved by this echo handler trick will be gone and gossiper's requirements get simplified, then we'd be able to turn off gossiper's echo pings and rely on our simpler failure detection.

**Comment by kostja** [comment]
> > Will gossip have to keep echo messages forever after this change?
> 
> Can't answer at this point.
> 
> Right now gossiper's failure detection has a bit different requirements than Raft's failure detection. For example, because of reasons (refs: [40b60e8](https://github.com/scylladb/scylladb/commit/40b60e8f09fda83248cba44374821a996ebdb036), #8013), gossiper wants to keep the replacing node marked as DOWN during the replace operation, which is done by the echo message handler returning an exception on the replacing node. On the other hand, we want to keep it simple and correctly detect servers if they are dead or alive when they join group 0, without special treatment for some servers and not others.
> 
> I can guess that in the future when we migrate all topology management to group 0, the problems solved by this echo handler trick will be gone and gossiper's requirements get simplified, then we'd be able to turn off gossiper's echo pings and rely on our simpler failure detection.

Gossiper or Raft failure detector cannot keep the node up or down. Their job is to collect accurate information about node state - and nodes having a unique identifier, so that this information is indeed correct (pertains to the right object).   It's clear the new failure detection method introduced in this patch is more accurate in this regard - since it identifies nodes by IDs, not by IPs, and ids are unique. But it is dropping generation numbers, so is disseminating less information than the previous version. My ask is can we (eventually) get the best of both worlds? And regarding the commit you're referring to, all topology operations, including repair, should at some point start using group0 state as their input, and get triggered via a centralized coordinator. Therefore, it's not  a concern (to me) that they depend on the echo information today - they should stop doing it eventually, and this can happen before gossip and raft FDs are merged.

**Comment by kbr-** [comment]
v2:
- in `cql_test_env` (used in boost tests), initialize `raft_group_registry` even later, because for some reason `system_keyspace` is available later in the startup procedure in `cql_test_env` than it is in `main`.

**Review by gleb-cloudius — APPROVED** [review]

**Comment by kbr-** [comment]
@tgrabiec please review.

But let's not merge yet (even if it's ready) - it conflicts a bit with https://github.com/scylladb/scylladb/pull/12035, I want to rebase on top of that when it's merged.

**Comment by avikivity** [comment]
If you add a verb, you have to add a cluster feature so you don't start using it prematurely.

**Comment by kbr-** [comment]
> If you add a verb, you have to add a cluster feature so you don't start using it prematurely.

It's already there: `SUPPORTS_RAFT_CLUSTER_MANAGEMENT`.

The direct failure detector sends RPCs only if endpoints are added to it, which happens only when members are added to group 0, which happens only if `SUPPORTS_RAFT_CLUSTER_MANAGEMENT` is enabled, which happens only when the RAFT experimental feature is enabled on every node in the cluster.

Matters would be more complex if we left experimental, but we didn't yet.

**Comment by kbr-** [comment]
Rebased onto `master` after https://github.com/scylladb/scylladb/pull/12035 was merged.
@tgrabiec please review.

**Comment by avikivity** [comment]
> > If you add a verb, you have to add a cluster feature so you don't start using it prematurely.
> 
> It's already there: `SUPPORTS_RAFT_CLUSTER_MANAGEMENT`.
> 
> The direct failure detector sends RPCs only if endpoints are added to it, which happens only when members are added to group 0, which happens only if `SUPPORTS_RAFT_CLUSTER_MANAGEMENT` is enabled, which happens only when the RAFT experimental feature is enabled on every node in the cluster.
> 
> Matters would be more complex if we left experimental, but we didn't yet.

I was referring to upgrades from 5.1. I guess we can sweep it under the experimental rug.

**Comment by kbr-** [comment]
> I was referring to upgrades from 5.1. I guess we can sweep it under the experimental rug.

We've made tons of other backward-incompatible changes, like changing definitions of data structures exchanged in RPCs etc. If you enabled Raft in 5.0/5.1, you won't be able to upgrade to 5.1/5.2.

**Comment by tgrabiec** [comment]
Major reason 2 would be also fixed by https://github.com/scylladb/scylladb/issues/6403.

File: service/raft/raft_group_registry.cc
Comment (tgrabiec) [inline service/raft/raft_group_registry.cc:272]: What about other raft verbs, don't we need them to reject verbs arriving at the wrong destination?

Comment (kbr-) [reply service/raft/raft_group_registry.cc:272]: True, we probably should.

It's interesting, because I haven't encountered issues with other verbs during replace. At a glance, it seems that (when replacing with the same IP) the replacing node should be handling RPCs intended for the replaced node, as well as RPCs intended for itself, which could lead to some weird situations. Or maybe Raft is already implemented in a way that handles all of this correctly?

In any case, the other Raft verbs are already passing `dst_id`, we only need to start checking them.

I'll create a follow-up issue.

cc @kostja

Comment (kbr-) [reply service/raft/raft_group_registry.cc:272]: Opened: https://github.com/scylladb/scylladb/issues/12179

**Comment by avikivity** [comment]
> We used GOSSIP_ECHO verb to perform failure detection. Now we use a special verb DIRECT_FD_PING introduced for this purpose.
> 
> There are multiple reasons to do so.
> 
> One minor reason: we want to use the same connection as other Raft verbs: if we can't deliver Raft append_entries or vote messages somewhere, that endpoint should be marked dead; if we can, the endpoint should be marked alive. So putting pings on the same connection as the other Raft verbs is important when dealing with weird situations where some connections are available but others are not. Observe that in `do_get_rpc_client_idx`, we put the new verb in the right place.


This introduces head-of-line blocking. If raft stuffs the connection with slow-to-process requests and the RPC server stops admitting them, then pings will be blocked behind them and the server will be needlessly marked dead.

**Comment by kbr-** [comment]
> This introduces head-of-line blocking. If raft stuffs the connection with slow-to-process requests and the RPC server stops admitting them, then pings will be blocked behind them and the server will be needlessly marked dead.

Whether the behavior is needless or actually desirable is not obvious.

This is the behavior of vanilla Raft, which uses append_entries RPCs to perform failure detection. append_entries are also used for replicating the log and updating the commit index. This mechanism is battle-tested in production.

Failure detection is mainly used for leader election. If the leader is so slow that a single append_entries (or ping in our case) cannot be exchanged in the period of one second, we probably do want to choose another leader. And that's what vanilla Raft would do.

**Comment by gleb-cloudius** [comment]
> > This introduces head-of-line blocking. If raft stuffs the connection with slow-to-process requests and the RPC server stops admitting them, then pings will be blocked behind them and the server will be needlessly marked dead.
> 
> Whether the behavior is needless or actually desirable is not obvious.
> 
> This is the behavior of vanilla Raft, which uses append_entries RPCs to perform failure detection. append_entries are also used for replicating the log and updating the commit index. This mechanism is battle-tested in production.
> 
> Failure detection is mainly used for leader election. If the leader is so slow that a single append_entries (or ping in our case) cannot be exchanged in the period of one second, we probably do want to choose another leader. And that's what vanilla Raft would do.

With vanilla raft each append_entries is also a keep alive message. In out case if the failure detector's ping is stuck behind a lot of append_entries a host will be declared as dead while in vanilla raft it would not (because it still gets append_entries, it is just ping that is stuck). I think we should feed append_entries into failure detector as well.

**Pull request closed without merging**
